### PR TITLE
[BugFix] fix wrong slot index when handling iceberg delete columns

### LINKED ISSUE #41866: BE crashes when access iceberg table with eq-delete files
### Steps to reproduce the behavior (Required)

**Create iceberg table with eq-delete files**

```
CREATE CATALOG hive_catalog WITH (
  'type'='iceberg',
  'catalog-type'='hive',
  'uri'='xxx',
  'clients'='5',
  'property-version'='1',
  'warehouse'='xxx'
);

use catalog hive_catalog;

use zya;

drop table flink_parquet;

CREATE TABLE flink_parquet(
  k1 int,
  k2 string,
  PRIMARY KEY(k1) NOT ENFORCED
) WITH (
    'format-version' = '2',
    'write.upsert.enabled' = 'true',
    'write-format' = 'parquet'
);

drop table flink_parquet2;

CREATE TABLE flink_parquet2(
  id STRING,
  yyyymmdd STRING,
  type STRING,
  PRIMARY KEY(id, yyyymmdd) NOT ENFORCED
) PARTITIONED BY (yyyymmdd)
WITH (
    'format-version' = '2',
    'write.upsert.enabled' = 'true',
    'write-format' = 'parquet'
);

insert into flink_parquet2(id, yyyymmdd, type) values('4', '2024-01-03', '5100'), ('5', '2024-01-02', '5555'), ('4', '2024-01-03', '3333');

insert into flink_parquet2(id, yyyymmdd, type) values('4', '2024-01-03', '5100'), ('5', '2024-01-02', '5555'), ('4', '2024-01-03', '3333');

```

**StarRocks query this iceberg table**

> select * from flink_parquet2;

### Expected behavior (Required)

FE returns "Unsupported iceberg file content:2", because we don't support parquet delete file yet.

However we don't expect BE crashes

### Real behavior (Required)

BE crashes

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`

main

### PR DESCRIPTION
**Pull request by dirtysalt**
## Why I'm doing:

We compute wrong slot index when dealing with deleted columns.

## What I'm doing:

data in `_materialize_index_in_chunk` should not exceed `slots.size()`

Fixes https://github.com/StarRocks/starrocks/issues/41866

When I'm debugging this issue, I can see we compute wrong slot index

**I0228 16:21:48.260135 490497 group_reader.cpp:398] [xxx] tuple desc = 0x6070000de9d0, slot desc = 0xbebebebebebebebe, slot index = 3**

```
I0228 16:21:46.591665 490466 descriptors.cpp:756] [xxx] create descriptor. tuple desc = 0x6070000de9d0, slot desc = 0x60d000050330, name = id
I0228 16:21:46.591701 490466 descriptors.cpp:756] [xxx] create descriptor. tuple desc = 0x6070000de9d0, slot desc = 0x60d000050400, name = yyyymmdd
I0228 16:21:46.591732 490466 descriptors.cpp:756] [xxx] create descriptor. tuple desc = 0x6070000de9d0, slot desc = 0x60d0000504d0, name = type
I0228 16:21:46.591763 490466 descriptors.cpp:756] [xxx] create descriptor. tuple desc = 0x6070000dea40, slot desc = 0x60d0000505a0, name = id
I0228 16:21:46.591782 490466 descriptors.cpp:756] [xxx] create descriptor. tuple desc = 0x6070000dea40, slot desc = 0x60d000050670, name = yyyymmdd
I0228 16:21:48.260039 490497 group_reader.cpp:53] [xxx] tuple desc = 0x6070000de9d0, slot desc = 0x60d000050330
I0228 16:21:48.260061 490497 group_reader.cpp:53] [xxx] tuple desc = 0x6070000de9d0, slot desc = 0x60d000050400
I0228 16:21:48.260088 490497 group_reader.cpp:53] [xxx] tuple desc = 0x6070000de9d0, slot desc = 0x60d0000504d0
I0228 16:21:48.260099 490497 group_reader.cpp:390] [xxx] tuple desc = 0x6070000de9d0, slot desc = 0x60d000050330
I0228 16:21:48.260105 490497 group_reader.cpp:390] [xxx] tuple desc = 0x6070000de9d0, slot desc = 0x60d000050400
I0228 16:21:48.260111 490497 group_reader.cpp:390] [xxx] tuple desc = 0x6070000de9d0, slot desc = 0x60d0000504d0
I0228 16:21:48.260121 490497 group_reader.cpp:398] [xxx] tuple desc = 0x6070000de9d0, slot desc = 0x60d000050330, slot index = 0
I0228 16:21:48.260128 490497 group_reader.cpp:398] [xxx] tuple desc = 0x6070000de9d0, slot desc = 0x60d0000504d0, slot index = 2
I0228 16:21:48.260135 490497 group_reader.cpp:398] [xxx] tuple desc = 0x6070000de9d0, slot desc = 0xbebebebebebebebe, slot index = 3
```

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/connector/hive_connector.cpp [hunk 1]
```diff
@@ -261,7 +264,13 @@ void HiveDataSource::_init_tuples_and_slots(RuntimeState* state) {
             _equality_delete_slots.emplace_back(d_slot_desc);
             if (!id_to_slots.contains(d_slot_desc->id())) {
                 _materialize_slots.push_back(d_slot_desc);
-                _materialize_index_in_chunk.push_back(delete_column_index++);
+                const auto& it = id_to_index.find(d_slot_desc->id());
+                if (it == id_to_index.end()) {
+                    return Status::InternalError(
+                            "Invalid slot id in delete_column_slot_ids. id =  " + std::to_string(d_slot_desc->id()) +
```
Comment (imay) [inline be/src/connector/hive_connector.cpp:270]: Could you use fmt::format instead?

Comment (imay) [reply be/src/connector/hive_connector.cpp:270]: And I want to suggest that change `_materialize_slots.push_back(d_slot_desc);` after this check succeeds.

**Review by imay — APPROVED** [review]

**Review by stephen-shelby — CHANGES_REQUESTED** [review]

File: be/src/connector/hive_connector.cpp [hunk 2]
```diff
@@ -251,17 +251,26 @@ void HiveDataSource::_init_tuples_and_slots(RuntimeState* state) {
...
 
-        int32_t delete_column_index = slots.size();
         auto* delete_column_tuple_desc =
                 state->desc_tbl().get_tuple_descriptor(_provider->_hdfs_scan_node.mor_tuple_id);
 
         std::vector<SlotDescriptor*> equality_delete_slots;
         for (SlotDescriptor* d_slot_desc : delete_column_tuple_desc->slots()) {
             _equality_delete_slots.emplace_back(d_slot_desc);
             if (!id_to_slots.contains(d_slot_desc->id())) {
+                const auto& it = id_to_index.find(d_slot_desc->id());
```
[the remark this answers is not in the document]
Comment (dirtysalt) [reply be/src/connector/hive_connector.cpp:266]: @stephen-shelby not exception, but error message. which is acceptable I guess.

[the remark this answers is not in the document]
Comment (stephen-shelby) [reply be/src/connector/hive_connector.cpp:266]: If you don't query the primary key column, how do id_to_index find the value?

**Review by stephen-shelby — APPROVED** [review]

**Comment by dirtysalt** [comment]
I guess there is a better to do that. see this commit https://github.com/StarRocks/starrocks/pull/42087/commits/d54f396f7a2a8ace966077141b71e614f82ff952

**Pull request closed without merging**